packages/api/src/external/carequality/document/process-document-retrieval-results.ts

```diff
@@ -0,0 +1,173 @@
+import { Bundle, BundleEntry } from "@medplum/fhirtypes";
+import { errorToString } from "@metriport/core/util/error/shared";
+import { capture } from "@metriport/core/util/notifications";
+import { DocumentReference } from "@metriport/ihe-gateway-sdk";
+import { MedicalDataSource } from "@metriport/core/external/index";
+import { DocumentRetrievalResult } from "../document-retrieval-result";
+import { isConvertible } from "../../fhir-converter/converter";
+import { setDocQueryProgressWithSource } from "../../hie/set-doc-query-progress-with-source";
+import { tallyDocQueryProgressWithSource } from "../../hie/tally-doc-query-progress-with-source";
+import { convertCDAToFHIR } from "../../fhir-converter/converter";
+import { upsertDocumentsToFHIRServer } from "../../fhir/document/save-document-reference";
+import { cqToFHIR } from "../../fhir/document";
+import { getDocuments } from "../../fhir/document/get-documents";
+import { ingestIntoSearchEngine } from "../../aws/opensearch";
+
+export async function processDocumentRetrievalResults({
+  requestId,
+  patientId,
+  cxId,
+  documentRetrievalResults,
+}: {
+  requestId: string;
+  patientId: string;
+  cxId: string;
+  documentRetrievalResults: DocumentRetrievalResult[];
+}): Promise<void> {
+  try {
+    for (const result of documentRetrievalResults) {
+      const { operationOutcome } = result.data;
+      const issuesWithGateway = operationOutcome?.issue?.length ?? 0;
+      const successDocsCount = result.data.documentReference?.length ?? 0;
+
+      tallyDocQueryProgressWithSource({
+        patient: { id: patientId, cxId: cxId },
+        progress: {
+          successful: successDocsCount,
+          errors: issuesWithGateway,
+        },
+        type: "download",
+        requestId,
+        source: MedicalDataSource.CAREQUALITY,
+      });
+
+      await handleDocReferences(result.data.documentReference, requestId, patientId, cxId);
+    }
+
+    await setDocQueryProgressWithSource({
+      patient: { id: patientId, cxId: cxId },
+      downloadProgress: {
+        status: "completed",
+      },
+      requestId,
+      source: MedicalDataSource.CAREQUALITY,
+    });
+  } catch (error) {
+    const msg = `Failed to process documents in Carequality.`;
+    console.log(`${msg}. Error: ${errorToString(error)}`);
+
+    await setDocQueryProgressWithSource({
+      patient: { id: patientId, cxId: cxId },
+      downloadProgress: { status: "failed" },
+      requestId,
+      source: MedicalDataSource.CAREQUALITY,
+    });
+
+    capture.message(msg, {
+      extra: {
+        context: `cq.processingDocuments`,
+        error,
+        patientId: patientId,
+        requestId,
+        cxId,
+      },
+      level: "error",
+    });
+    throw error;
+  }
+}
+
+async function handleDocReferences(
+  docRefs: DocumentReference[],
+  requestId: string,
+  patientId: string,
+  cxId: string
+) {
+  // REMINDER: YOU NEED TO UPDATE MIRTH TO SEND THE ABOVE
```

Comment:
In the process of getting mirth to send bucketname, filename, isNew. Will remove when done